python/tvm/relax/transform/optimize_layout_transform.py

```diff
@@ -63,27 +63,24 @@ def transform_function(self, func: Expr, mod: IRModule, ctx: PassContext) -> IRM
 
         self.mod = mod
         updated_func = func
-        for _, func in mod.functions_items():
-            # Skip non-relax functions
-            if not isinstance(func, Function):
-                continue
-            # Skip primitive functions
-            if "Primitive" in func.attrs.keys() and func.attrs["Primitive"] != 0:
-                continue
 
-            def rewriter(expr, matches):
-                arg1 = matches[self.pattern]
-                if self.pattern_2 not in matches.keys():
+        # Skip primitive functions
+        if "Primitive" in func.attrs.keys() and func.attrs["Primitive"] != 0:
+            return updated_func
+
+        def rewriter(expr, matches):
+            arg1 = matches[self.pattern]
+            if self.pattern_2 not in matches.keys():
+                arg2 = matches[self.input]
+            else:
+                arg2 = matches[self.gv_]
+                if "remove_pad" == self.mod[arg2].attrs["operator_name"]:
                     arg2 = matches[self.input]
-                else:
-                    arg2 = matches[self.gv_]
-                    if "remove_pad" == self.mod[arg2].attrs["operator_name"]:
-                        arg2 = matches[self.input]
-                if hasattr(arg1.struct_info, "shape") and hasattr(arg2.struct_info, "shape"):
-                    if structural_equal(arg1.struct_info.shape, arg2.struct_info.shape):
-                        return arg2
-                return expr
+            if hasattr(arg1.struct_info, "shape") and hasattr(arg2.struct_info, "shape"):
+                if structural_equal(arg1.struct_info.shape, arg2.struct_info.shape):
+                    return arg2
+            return expr
 
-            updated_func = rewrite_call(self.pattern, rewriter, func)
```

Comment:
Beyond just repetition, it looks like the previous implementation would produce incorrect results for an `IRModule` that contains more than one non-primitive Relax function.  Rather than returning the updated version of the argument `func`, this would return the updated version of whichever non-primitive relax function occurs last in `mod.function_items()`, regardless of the `func` argument.

Reply:
Thanks for the review, don't know how I missed that.